File: CS/TreeListCellMerging/MyTreeListOperationPrintEachNode.cs
using DevExpress.XtraPrinting;
using DevExpress.XtraTreeList;
using DevExpress.XtraTreeList.Nodes;
using DevExpress.XtraTreeList.Nodes.Operations;
using DevExpress.XtraTreeList.Printing;
using DevExpress.XtraTreeList.ViewInfo;
using System.Drawing;

namespace TreeListCellMerging
{
    public class MyTreeListOperationPrintEachNode : TreeListOperationPrintEachNode
    {
        CellInfo prevCell = null;
        bool printAllNodes;
        TreeList treeList;

        public MyTreeListOperationPrintEachNode(TreeList treeList, TreeListPrinter printer, TreeListViewInfo viewInfo, bool printTree, bool printImages, bool printCheckBoxes, bool printAllNodes)
            : base(treeList, printer, viewInfo, printTree, printImages, printCheckBoxes)
        {
            this.treeList = treeList;
            this.printAllNodes = printAllNodes;
        }

        protected override VisualBrick CreateCellBrick(CellInfo cell, DevExpress.XtraTreeList.Nodes.TreeListNode node)
        {
            VisualBrick brick = base.CreateCellBrick(cell, node) as VisualBrick;

            int lastColumnIndex = treeList.Columns.Count - 1;
            int prevIndex = lastColumnIndex - 1;
            if(node.GetDisplayText(lastColumnIndex) == node.GetDisplayText(prevIndex))
            {
                if(cell.Column.AbsoluteIndex == lastColumnIndex)
                {
                    brick.Sides = BorderSide.Right | BorderSide.Bottom | BorderSide.Top;
                    Rectangle rect = cell.EditorViewInfo.Bounds;
                    rect.X -= prevCell.EditorViewInfo.Bounds.Width;
                    rect.Width += prevCell.EditorViewInfo.Bounds.Width;
                    rect.Inflate(1, 1);
                    cell.SetBounds(rect, new System.Windows.Forms.Padding(0));
                }

                else if(cell.Column.AbsoluteIndex == lastColumnIndex - 1)
                {
                    brick.Sides = BorderSide.Left | BorderSide.Bottom | BorderSide.Top;
                    (brick as TextBrick).HorzAlignment = DevExpress.Utils.HorzAlignment.Far;
                    prevCell = cell;
                }
            }
            return brick;
        }

        public override bool NeedsVisitChildren(TreeListNode node) { return node.Expanded || printAllNodes; }
    }
}

File: VB/TreeListCellMerging/MyTreeListOperationPrintEachNode.vb
﻿Imports Microsoft.VisualBasic
Imports DevExpress.XtraPrinting
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Nodes
Imports DevExpress.XtraTreeList.Nodes.Operations
Imports DevExpress.XtraTreeList.Printing
Imports DevExpress.XtraTreeList.ViewInfo
Imports System.Drawing

Namespace TreeListCellMerging
	Public Class MyTreeListOperationPrintEachNode
		Inherits TreeListOperationPrintEachNode
		Private prevCell As CellInfo = Nothing
		Private printAllNodes As Boolean
		Private treeList As TreeList

		Public Sub New(ByVal treeList As TreeList, ByVal printer As TreeListPrinter, ByVal viewInfo As TreeListViewInfo, ByVal printTree As Boolean, ByVal printImages As Boolean, ByVal printCheckBoxes As Boolean, ByVal printAllNodes As Boolean)
			MyBase.New(treeList, printer, viewInfo, printTree, printImages, printCheckBoxes)
			Me.treeList = treeList
			Me.printAllNodes = printAllNodes
		End Sub

		Protected Overrides Function CreateCellBrick(ByVal cell As CellInfo, ByVal node As DevExpress.XtraTreeList.Nodes.TreeListNode) As VisualBrick
			Dim brick As VisualBrick = TryCast(MyBase.CreateCellBrick(cell, node), VisualBrick)

			Dim lastColumnIndex As Integer = treeList.Columns.Count - 1
			Dim prevIndex As Integer = lastColumnIndex - 1
			If node.GetDisplayText(lastColumnIndex) = node.GetDisplayText(prevIndex) Then
				If cell.Column.AbsoluteIndex = lastColumnIndex Then
					brick.Sides = BorderSide.Right Or BorderSide.Bottom Or BorderSide.Top
					Dim rect As Rectangle = cell.EditorViewInfo.Bounds
					rect.X -= prevCell.EditorViewInfo.Bounds.Width
					rect.Width += prevCell.EditorViewInfo.Bounds.Width
					rect.Inflate(1, 1)
					cell.SetBounds(rect, New System.Windows.Forms.Padding(0))

				ElseIf cell.Column.AbsoluteIndex = lastColumnIndex - 1 Then
					brick.Sides = BorderSide.Left Or BorderSide.Bottom Or BorderSide.Top
					TryCast(brick, TextBrick).HorzAlignment = DevExpress.Utils.HorzAlignment.Far
					prevCell = cell
				End If
			End If
			Return brick
		End Function

		Public Overrides Function NeedsVisitChildren(ByVal node As TreeListNode) As Boolean
			Return node.Expanded OrElse printAllNodes
		End Function
	End Class
End Namespace
